File: 146-lru-cache/lru-cache.py
class DoublyLinkedList:
    def __init__(self,key=None, value=None):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None
class LRUCache:

    def __init__(self, capacity: int):
        self.head = DoublyLinkedList()
        self.tail = DoublyLinkedList()
        self.head.prev = self.tail
        self.tail.next = self.head 
        self.capacity = capacity
        self.node_map = {}

    def delete(self, node):
        if node in [self.head, self.tail]:
            return
        del self.node_map[node.key]
        node.prev.next = node.next
        node.next.prev = node.prev
    
    def insert(self, node):
        self.node_map[node.key] = node
        node.next = self.head
        node.prev = self.head.prev
        self.head.prev.next = node
        self.head.prev = node

    def get(self, key: int) -> int:
        if key not in self.node_map:
            return -1
        node = self.node_map[key]
        self.delete(node)
        self.insert(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        node = DoublyLinkedList(key, value)
        if key in self.node_map:
            self.delete(self.node_map[key])
        if self.capacity == len(self.node_map):
            self.delete(self.tail.next)
        self.insert(node)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146-lru-cache/lru-cache.java
class DoublyLinkedNode {
    public DoublyLinkedNode prev;
    public DoublyLinkedNode next;
    public int key;
    public int val;

    public DoublyLinkedNode() {

    }
}

class LRUCache {
    private int capacity;
    private DoublyLinkedNode head;
    private DoublyLinkedNode tail;
    private Map<Integer, DoublyLinkedNode> nodeMap;
    public LRUCache(int capacity) {
        this.capacity = capacity;
        this.head = new DoublyLinkedNode();
        this.tail = new DoublyLinkedNode();
        this.head.prev = this.tail;
        this.tail.next = this.head;
        this.nodeMap = new HashMap<>();
        
    }
    
    public int get(int key) {
        if (!nodeMap.containsKey(key)) {
            return  -1;
        } else {
            DoublyLinkedNode node = nodeMap.get(key);
            remove(key);
            insert(key, node);
            return nodeMap.get(key).val;
        }
    }

    public void insert(int key, DoublyLinkedNode node) {
        node.next = head;
        node.prev = head.prev;
        head.prev.next = node;
        head.prev = node;
        nodeMap.put(key, node);
    }
    public void remove(int key) {
        DoublyLinkedNode node = nodeMap.get(key);
        // System.out.println(key);
        // System.out.println(node.val);
        node.prev.next = node.next;
        node.next.prev = node.prev;
        nodeMap.remove(key);
    }
    
    public void put(int key, int value) {
        if (nodeMap.containsKey(key)){
            remove(key);
        } else if (nodeMap.size() == capacity) {
            remove(tail.next.key);
        }
        DoublyLinkedNode node = new DoublyLinkedNode();
        node.val = value;
        node.key = key;
        insert(key, node);
        //System.out.println(nodeMap);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */